pyunlocbox/acceleration.py

```diff
@@ -0,0 +1,502 @@
+                    return np.sum([f.eval(x) for f in self.functions])
+                # Solution at previous extrapolation
+                xk = self.buffer[0]
+                # Search direction
+                pk = extrap - xk
+                # Objective value during the previous extrapolation
+                old_fval = np.sum(objective[-self.k])
+
+                a, fc, fa = line_search_armijo(f=f,
+                                               xk=xk,
+                                               pk=pk,
+                                               gfk=pk,
+                                               old_fval=old_fval,
+                                               c1=1e-4,
+                                               alpha0=1.)
+
+                # New point proposal
+                if a is None:
+                    warnings.warn('Line search failed to find good step size')
+                else:
+                    extrap[:] = xk + a * pk
+
+                logging.debug('extrap = {}'.format(extrap))
+
+            # Clear buffer and parameter grid for next extrapolation process
+            self.buffer = []
+            self.lambda_ = [] if self.adaptive else self.lambda_
+
+            return extrap
+
+        else:  # Gather points for future extrapolation
+            self.buffer.append(copy.copy(solver.sol))
+            return solver.sol
+
+    def _post(self):
+        del self.buffer, self.functions
+
+# -----------------------------------------------------------------------------
+# Mixed optimizers
+# -----------------------------------------------------------------------------
+
+
+class fista_backtracking(backtracking, fista):
+    r"""
+    Acceleration scheme with backtracking for forward-backward solvers.
+
+    Notes
+    -----
+    This is the acceleration scheme and backtracking strategy proposed in the
+    original FISTA paper, :cite:`beck2009FISTA`.
+
+    Examples
+    --------
+    >>> from pyunlocbox import functions, solvers, acceleration
+    >>> import numpy as np
+    >>> y = [4, 5, 6, 7]
+    >>> x0 = np.zeros(len(y))
+    >>> f1 = functions.norm_l2(y=y)
+    >>> f2 = functions.dummy()
+    >>> accel=acceleration.fista_backtracking()
+    >>> solver = solvers.forward_backward(accel=accel, step=0.5)
+    >>> ret = solvers.solve([f1, f2], x0, solver, atol=1e-5)
+    Solution found after 13 iterations:
+        objective function f(sol) = 9.518528e-08
+        stopping criterion: ATOL
+    >>> ret['sol']
+    array([ 3.99989006,  4.99986257,  5.99983509,  6.9998076 ])
+
+    """
+
+    def __init__(self, **kwargs):
+        """
+        I can do multiple inheritance here and avoid the deadly diamond of
+        death because the classes backtracking and fista modify different
+        methods of their parent class dummy. If that would not be the case, I
+        guess the best solution would be to inherit from accel and rewrite the
+        _update_step() and _update_sol() methods.
+        """
```

Comment:
I looked a bit into ABC documentation and tutorials, and it seems to me that this mechanism would be ideal for defining the `pyunlocbox.solver` and `pyunlocbox.accel` base classes, but I fail to see how it would help here for generating this mixed zfista_backtracking` class. In any case, I think it's worth it to adapt `pyunlocbox.solver` and `pyunlocbox.accel` into abstract base classes, but I would delegate that to a future release, as we would have to ensure that normal behavior doesn't change when doing these modifications.